sqlmesh/core/engine_adapter/base.py

```diff
@@ -1208,23 +1267,131 @@ def scd_type_2(
+        # column names and then remove them from the unmanaged_columns
+        if check_columns and check_columns == exp.Star():
+            check_columns = [exp.column(col) for col in unmanaged_columns]
+        execution_ts = self._to_utc_timestamp(to_ts(execution_time), time_data_type)
+        update_valid_from_start: t.Union[str, exp.Expression]
+        if updated_at_as_valid_from:
+            if not updated_at_name:
+                raise SQLMeshError(
+                    "Cannot use `updated_at_as_valid_from` without `updated_at_name` for SCD Type 2"
+                )
+            update_valid_from_start = updated_at_name
+        elif execution_time_as_valid_from:
+            update_valid_from_start = execution_ts
+        else:
+            update_valid_from_start = self._to_utc_timestamp(
+                "1970-01-01 00:00:00+00:00", time_data_type
+            )
+        insert_valid_from_start = execution_ts if check_columns else exp.to_column(updated_at_name)  # type: ignore
+        if check_columns:
+            row_check_conditions = []
+            for col in check_columns:
+                t_col = col.copy()
+                t_col.this.set("this", f"t_{col.name}")
+                row_check_conditions.extend(
+                    [
+                        col.neq(t_col),
+                        exp.and_(t_col.is_(exp.Null()), col.is_(exp.Null()).not_()),
+                        exp.and_(t_col.is_(exp.Null()).not_(), col.is_(exp.Null())),
+                    ]
+                )
+            row_value_check = exp.or_(*row_check_conditions)
+            unique_key_conditions = []
+            for col in unique_key:
+                t_col = col.copy()
+                t_col.this.set("this", f"t_{col.name}")
+                unique_key_conditions.extend(
+                    [t_col.is_(exp.Null()).not_(), col.is_(exp.Null()).not_()]
+                )
+            unique_key_check = exp.and_(*unique_key_conditions)
+            # unique_key_check is saying "if the row is updated"
+            # row_value_check is saying "if the row has changed"
+            updated_row_filter = exp.and_(unique_key_check, row_value_check)
+            # joined._exists IS NULL is saying "if the row is deleted"
+            valid_to_case_stmt = (
+                exp.Case()
+                .when(
+                    exp.or_(
+                        exp.column("_exists", "joined").is_(exp.Null()),
+                        updated_row_filter,
+                    ),
+                    execution_ts,
+                )
+                .else_(exp.to_column(f"t_{valid_to_name}"))
+                .as_(valid_to_name)
+            )
+            valid_from_case_stmt = exp.func(
+                "COALESCE",
+                exp.column(f"t_{valid_from_name}"),
+                update_valid_from_start,
+            ).as_(valid_from_name)
+        else:
+            assert updated_at_name is not None
```

Comment:
is this just for mypy? could be simplified to assert updated_at_name

Reply:
or why not make 1365 elif updated_at_name?

Reply:
Yes it is for mypy and it is impossible to hit this and not have an `uppdated_at_name`. The public methods that are used to call this private method ensure that is true.

Reply:
Well technically possible if someone called the private method directly but impossible in practice since the evaluator only calls public methods.